WcaOnRails/app/models/result.rb

```diff
@@ -1,15 +1,82 @@
 class Result < ActiveRecord::Base
+  include ResultMethods
+
   self.table_name = "Results"
 
   belongs_to :competition, foreign_key: :competitionId
+  belongs_to :country, foreign_key: :countryId
   belongs_to :person, -> { current }, primary_key: :wca_id, foreign_key: :personId
   belongs_to :round, foreign_key: :roundId
   belongs_to :event, foreign_key: :eventId
+  belongs_to :format, foreign_key: :formatId
 
   scope :podium, -> { joins(:round).merge(Round.final_rounds).where(pos: [1..3]).where("best > 0") }
   scope :winners, -> { joins(:round, :event).merge(Round.final_rounds).where("pos = 1 and best > 0").order("Events.rank") }
 
+  validate :best_and_average
+  def best_and_average
+    return errors.add(:competitionId, "invalid") unless competition
+    return errors.add(:countryId, "invalid") unless country
+    return errors.add(:roundId, "invalid") unless round
+    return errors.add(:eventId, "invalid") unless event
+    return errors.add(:formatId, "invalid") unless format
+    return errors.add(:base, "Cannot skip all solves.") if solve_times.all?(&:skipped?)
+
+    solve_times.each_with_index do |solve_time, i|
+      unless solve_time.valid?
+        errors.add(:"value#{i + 1}", solve_time.errors.messages[:base].join(" "))
+      end
+    end
+
+    previous_solve_time = solve_times.first
+    solve_times[1..-1].each do |solve_time|
+      if !solve_time.skipped? && previous_solve_time.skipped?
+        errors.add(:base, "Skipped solves must all come at the end.")
+        return
+      end
+      previous_solve_time = solve_time
+    end
+
+    unskipped_count = solve_times.length - solve_times.count(&:skipped?)
+    if round.combined?
+      if unskipped_count > format.expected_solve_count
+        errors.add(:base, "Expected at most #{format.expected_solve_count} #{'solve'.pluralize(format.expected_solve_count)}, but found #{unskipped_count}.")
+        return
+      end
+    else
+      if unskipped_count != format.expected_solve_count
+        errors.add(:base, "Expected #{format.expected_solve_count} #{'solve'.pluralize(format.expected_solve_count)}, but found #{unskipped_count}.")
+        return
+      end
+    end
+
+    correct_best_solve_time = sorted_solves_with_index.first[0]
+    if correct_best_solve_time.wca_value != best
+      errors.add(:best, "should be #{correct_best_solve_time.wca_value}")
+      return
+    end
+
+    if eventId == "333fm"
+      sum_moves = 0
+      counting_solve_times.each do |solve_time|
+        sum_moves += solve_time.move_count
+      end
+      correct_average_wca_value = 100 * sum_moves / counting_solve_times.length
+    else
+      sum_centis = 0
+      counting_solve_times.each do |solve_time|
+        sum_centis += solve_time.time_centiseconds
+      end
+      correct_average_wca_value = sum_centis / counting_solve_times.length
+    end
+
+    if correct_average_wca_value != average
+      errors.add(:average, "should be #{correct_average_wca_value}")
+      return
+    end
```

Comment:
In the earlier comment I meant - but didn't say - that i like these patterns applied to the whole function

`return errors.add(...)`
`return errors.add(...) if condition`